lib/src/analysis_options/analysis_options_file.dart

```diff
@@ -0,0 +1,116 @@
+// BSD-style license that can be found in the LICENSE file.
+import 'package:yaml/yaml.dart';
+
+import 'file_system.dart';
+import 'merge_options.dart';
+
+/// The analysis options configuration is a dynamically-typed JSON-like data
+/// structure.
+///
+/// (It's JSON-*like* and not JSON because maps in it may have non-string keys.)
+typedef AnalysisOptions = Map<Object?, Object?>;
+
+/// Interface for taking a "package:" URI that may appear in an analysis
+/// options file's "include" key and resolving it to a file path which can be
+/// passed to [FileSystem.join()].
+typedef ResolvePackageUri = Future<String?> Function(Uri packageUri);
+
+/// Reads an `analysis_options.yaml` file in [directory] or in the nearest
+/// surrounding folder that contains that file using [fileSystem].
+///
+/// Stops walking parent directories as soon as it finds one that contains an
+/// `analysis_options.yaml` file. If it reaches the root directory without
+/// finding one, returns an empty [YamlMap].
+///
+/// If an `analysis_options.yaml` file is found, reads it and parses it to a
+/// [YamlMap]. If the map contains an `include` key whose value is a list, then
+/// reads any of the other referenced YAML files and merges them into this one.
+/// Returns the resulting map with the `include` key removed.
+///
+/// If there any "package:" includes, then they are resolved to file paths
+/// using [resolvePackageUri]. If [resolvePackageUri] is omitted, an exception
+/// is thrown if any "package:" includes are found.
+Future<AnalysisOptions> findAnalysisOptions(
+    FileSystem fileSystem, FileSystemPath directory,
+    {ResolvePackageUri? resolvePackageUri}) async {
+  while (true) {
+    var optionsPath = await fileSystem.join(directory, 'analysis_options.yaml');
+    if (await fileSystem.fileExists(optionsPath)) {
+      return readAnalysisOptions(fileSystem, optionsPath,
+          resolvePackageUri: resolvePackageUri);
+    }
+
+    var parent = await fileSystem.parentDirectory(directory);
+    if (parent == null) break;
+    directory = parent;
+  }
+
+  // If we get here, we didn't find an analysis_options.yaml.
+  return const {};
+}
+
+/// Uses [fileSystem] to read the analysis options file at [optionsPath].
+///
+/// If there any "package:" includes, then they are resolved to file paths
+/// using [resolvePackageUri]. If [resolvePackageUri] is omitted, an exception
+/// is thrown if any "package:" includes are found.
+Future<AnalysisOptions> readAnalysisOptions(
+        FileSystem fileSystem, FileSystemPath optionsPath,
+        {ResolvePackageUri? resolvePackageUri}) async =>
+    _readAnalysisOptions(fileSystem, resolvePackageUri, optionsPath);
+
+Future<AnalysisOptions> _readAnalysisOptions(FileSystem fileSystem,
+    ResolvePackageUri? resolvePackageUri, FileSystemPath optionsPath) async {
+  var yaml = loadYamlNode(await fileSystem.readFile(optionsPath));
+
+  // Lower the YAML to a regular map.
+  if (yaml is! YamlMap) return const {};
```

Comment:
[nit] Returning an empty map doesn't correspond to the comment above. Can you move the comment below this line, and consider adding a comment explaining why reading a non-map defaults to empty options.